pyproject.toml

```diff
@@ -0,0 +1,9 @@
+[build-system]
+requires = [
+    "setuptools>=42",
+    "wheel",
+    "numpy>=1.16.6",
+    "scipy>=1.0",
+    "qutip>=4.6",
```

Comment:
If you have no code that actually executes or runs in a `setup.py`, do you need `numpy`, `scipy` or `qutip` at build time?

Reply:
I'm always a bit confused by what should be in `requirements.txt `and `pyporject.toml`.

So `pyproject.toml` is used by `pip` to build the project when installing. Without `setup.py`, none of the python script in `qutip_qip` is executed so that I don't have any dependencies for the build except for the build tools.

And `requirements.txt` includes things needed for running the code.

That roughly correct?

Reply:
`requirements.txt` is basically nothing,  It has no formal part of the build/installation process at all - I think maybe it was used more in the days before the dependency manager was reliable?  I use `requirements.txt` in the documentation build to exactly specify a complete environment to do the builds in; it inclues exact pinned versions of every package I had installed when I ran a successful documentation build, similar to a `package-lock.json` file in the npm world.  You don't even need to have one if it's not appropriate.

The `build-system.requires` key in `pyproject.toml` has everything that needs to be installed to create the sdists and wheels - if you don't actually execute any Python code/do any compilation, you're probably fine with `setuptools` and `wheel` in there only.  This key deprecates the `setup_requires` key in `setup.cfg` or `setup.py` if you're using `pip install -e .` to get a development version; pip will read from `pyproject.toml` _before_ setuptools would have a chance to see what was going on.  If you're still using `setup.py develop` to get an editable build (useful in QuTiP because skipping pip's venvs makes Cython development a little more convenient), then setuptools won't see `pyproject.toml` (because you ran `setup.py` directly), and having the version requirements also available there _can_ be a useful check.  For you, you just need `pyproject.toml`.

The `install_requires` key in `setup.cfg` is the "classic" dependencies.  This is what a user actually needs to have installed for your package to work at run-time.

Reply:
Historical note: `requirements.txt` was in vogue for awhile some years ago because it provided a means to provide dependencies in a way that gave access to all of pip's capabilities and didn't require executing potentially complex Python code (e.g. `setup.py`). It's now been superseded by `pyproject.toml` although it's still useful when one wants to provide a list of dependencies without any other functionality for some reason (e.g. a `requirements-doc.txt` might specify the packages needed to build documentation, but not be part of any bigger packaging process).